src/webui/www/private/scripts/client.js

```diff
@@ -453,33 +452,54 @@ window.addEventListener("DOMContentLoaded", () => {
     };
 
     const updateCategoryList = function() {
-        const categoryList = $("categoryFilterList");
+        const categoryList = document.getElementById("categoryFilterList");
         if (!categoryList)
             return;
         categoryList.getChildren().each(c => c.destroy());
 
         const categoryItemTemplate = document.getElementById("categoryFilterItem");
 
-        const create_link = function(hash, text, count) {
-            let display_name = text;
-            let margin_left = 0;
-            if (useSubcategories) {
-                const category_path = text.split("/");
-                display_name = category_path[category_path.length - 1];
-                margin_left = (category_path.length - 1) * 20;
-            }
-
+        const createCategoryLink = (hash, name, count) => {
             const categoryFilterItem = categoryItemTemplate.content.cloneNode(true).firstElementChild;
             categoryFilterItem.id = hash;
             categoryFilterItem.classList.toggle("selectedFilter", hash === selectedCategory);
 
             const span = categoryFilterItem.firstElementChild;
-            span.style.marginLeft = `${margin_left}px`;
-            span.lastChild.textContent = `${display_name} (${count})`;
+            span.lastElementChild.textContent = `${name} (${count})`;
 
             return categoryFilterItem;
         };
 
+        const createCategoryTree = (category, root = categoryList) => {
+            const displayName = category.segments.at(-1);
+            const li = createCategoryLink(category.categoryHash, displayName, category.categoryCount);
+            li.firstElementChild.style.paddingLeft = `${(category.segments.length - 1) * 20 + 6}px`;
+
+            if (category.parentID === null)
+                categoriesFragment.appendChild(li);
+            else
+                root.appendChild(li);
+
+            if (category.children.length === 0) {
+                // Expand ancestor categories after creating subcategory
+                if (LocalPreferences.get(`category_${category.categoryHash}_collapsed`) === null) {
+                    LocalPreferences.set(`category_${category.categoryHash}_collapsed`, "false");
+                    return true;
+                }
+                return false;
+            }
+
+            li.querySelector(".categoryToggle").style.visibility = "visible";
+            const ul = document.createElement("ul");
+            li.appendChild(ul);
+
+            const forceExpand = category.children.reduce((acc, subcategory) => createCategoryTree(subcategory, ul) || acc, false);
```

Comment:
I chose recursive approach because it was simply easier to write it that way, given tree-like structure of the data. Since we are dealing with only one tree at a time, I thought that we should always be within reasonable recursion depth (unless somebody goes insanely crazy with the number of categories...). I'm not against it so if you are adamant about it, I'll do it.

Reply:
I think moving to for loop would be safer, since there are some variables used in the recursive function and those will take space on the stack and effectively lower the max recursive depth.
https://stackoverflow.com/questions/34570551/call-stack-size-in-recursive-functions-maximum-call-stack-size-is-lower-than-ex

Reply:
Done as requested 👍